Sources/StreamChatUI/Navigation/NavigationRouter.swift

```diff
@@ -0,0 +1,34 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import UIKit
+
+/// A root class for all routes in the SDK.
+///
+/// Router objects are used to handle navigation between view controllers.
+///
+/// - Important: ⚠️ The lifetime of a router must be similar to the lifetime of its `rootViewController`. Only the
+/// root view controller is allowed to hold a strong reference to its router object.
+///
+open class NavigationRouter<Controller: UIViewController>: UIResponder {
+    /// The root `UIViewController` object of this router.
+    public unowned var rootViewController: Controller
```

Comment:
Btw why do we need to know the concrete type of the rootViewController? 🤔 That just makes it dangerous that the Router can call actions on the view controller, and also limits the router in a way that it can only be launched by the provided controller, where if it was just `UIViewController`, it can be launched from everywhere, even from a custom view controller provided by the customer. 

And couldn't `NavigationRouter`just be a protocol? 🤔

Reply:
> Btw why do we need to know the concrete type of the rootViewController?
I was asking the same question and at the moment I don't have a good answer for that. I wasn't sure if it's a good idea to remove it or not. Let me think of it a bit more, and I'll remove it if I don't find a good reason for keeping it.

> And couldn't NavigationRouterjust be a protocol? 🤔
It could be. However, it's easier to have it like this because you want to make sure `rootViewController` is referenced weakly.

Reply:
@nuno-vieira , so it is needed and it's quite handy. It allows us to reason about the type of the root VC. For example, in `showMessageList` we get the current client instance from the root VC. So let's keep it 👍